File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/AdapterViewItemSelectionObservable.java
package com.jakewharton.rxbinding2.widget;

import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static android.widget.AdapterView.INVALID_POSITION;
import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class AdapterViewItemSelectionObservable extends InitialValueObservable<Integer> {
  private final AdapterView<?> view;

  AdapterViewItemSelectionObservable(AdapterView<?> view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super Integer> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    view.setOnItemSelectedListener(listener);
    observer.onSubscribe(listener);
  }

  @Override protected Integer getInitialValue() {
    return view.getSelectedItemPosition();
  }

  static final class Listener extends MainThreadDisposable implements OnItemSelectedListener {
    private final AdapterView<?> view;
    private final Observer<? super Integer> observer;

    Listener(AdapterView<?> view, Observer<? super Integer> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int position, long id) {
      if (!isDisposed()) {
        observer.onNext(position);
      }
    }

    @Override public void onNothingSelected(AdapterView<?> adapterView) {
      if (!isDisposed()) {
        observer.onNext(INVALID_POSITION);
      }
    }

    @Override protected void onDispose() {
      view.setOnItemSelectedListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding3/widget/AdapterViewItemSelectionObservable.kt
@file:JvmName("RxAdapterView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.widget

import android.view.View
import android.widget.Adapter
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import com.jakewharton.rxbinding3.InitialValueObservable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import android.widget.AdapterView.INVALID_POSITION
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of the selected position of `view`. If nothing is selected,
 * [AdapterView.INVALID_POSITION] will be emitted.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun <T : Adapter> AdapterView<T>.itemSelections(): InitialValueObservable<Int> {
  return AdapterViewItemSelectionObservable(this)
}

private class AdapterViewItemSelectionObservable(
  private val view: AdapterView<*>
) : InitialValueObservable<Int>() {

  override fun subscribeListener(observer: Observer<in Int>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    view.onItemSelectedListener = listener
    observer.onSubscribe(listener)
  }

  override val initialValue get() = view.selectedItemPosition

  private class Listener(
    private val view: AdapterView<*>,
    private val observer: Observer<in Int>
  ) : MainThreadDisposable(), OnItemSelectedListener {

    override fun onItemSelected(adapterView: AdapterView<*>, view: View?, position: Int, id: Long) {
      if (!isDisposed) {
        observer.onNext(position)
      }
    }

    override fun onNothingSelected(adapterView: AdapterView<*>) {
      if (!isDisposed) {
        observer.onNext(INVALID_POSITION)
      }
    }

    override fun onDispose() {
      view.onItemSelectedListener = null
    }
  }
}
